### ISSUE TITLE
Discussion

### ISSUE THREAD

**Issue by YipingNUS** [AdenosHermes/pytorch-NLPTutorial#1]
Hi, 

Thanks for posting the solution. I'm a beginner with pyTorch and just spent the afternoon working out the solution. A few comments below:

1. Based on my understanding of the tutorial, the character level model doesn't care about the POS tags. It just encodes the char sequence to a hidden vector, which is used by the final model to predict POS tag. So the forward method of CharEncoder can simply return the last hidden layer. I have the following:

```python
    def forward(self, word):
        embeds = self.char_embeddings(word)
        lstm_out, self.hidden = self.lstm(
            embeds.view(len(word), 1, -1), self.hidden)
        return lstm_out[-1].view(1,-1)
```

In my main method I have the following:

```python
for epoch in range(300):  # again, normally you would NOT do 300 epochs, it is toy data
    for sentence, tags in training_data:
        # Step 1. Remember that Pytorch accumulates gradients.
        # We need to clear them out before each instance
        model.zero_grad()
        char_encoder.zero_grad()
        
        # Also, we need to clear out the hidden state of the LSTM,
        # detaching it from its history on the last instance.
        model.hidden = model.init_hidden()

        # Step 2. Get our inputs ready for the network, that is, turn them into
        # Tensors of word indices.
        sentence_in = prepare_sequence(sentence, word_to_ix)

        # Step 2.a encode char sequence
        char_encodings = []
        for word in sentence:
            char_encoder.hidden = char_encoder.init_hidden()
            char_in = prepare_sequence(word, char_to_ix)
            char_encoding = char_encoder(char_in)
            char_encodings.append(char_encoding)
        
        char_encodings = torch.cat(char_encodings)
        targets = prepare_sequence(tags, tag_to_ix)

        # Step 3. Run our forward pass.
        tag_scores = model(sentence_in, char_encodings)

        # Step 4. Compute the loss, gradients, and update the parameters by
        #  calling optimizer.step()
        loss = loss_function(tag_scores, targets)
        loss.backward()
        optimizer.step()
```

The key differences are:

1. I initialise the hidden layer for char_encoder for every word. That means I encode each word regardless of the context. I think either way are acceptable.
2. I call back prop only for the final classifier. I think the gradient will be propagated to the char level lstm.

I might be wrong since I've very limited experience with pytorch.
